naga/src/back/hlsl/help.rs

```diff
@@ -1031,6 +1051,208 @@ impl<W: Write> super::Writer<'_, W> {
                         // End of function body
                         writeln!(self.out, "}}")?;
                     }
+                    crate::MathFunction::Abs
+                        if matches!(
+                            func_ctx.resolve_type(arg, &module.types).scalar(),
+                            Some(crate::Scalar {
+                                kind: crate::ScalarKind::Sint,
+                                width: 4,
+                            })
+                        ) =>
+                    {
+                        let arg_ty = func_ctx.resolve_type(arg, &module.types);
+                        let scalar = arg_ty.scalar().unwrap();
+                        let components = arg_ty.components();
+
+                        let wrapped = WrappedMath {
+                            fun,
+                            scalar,
+                            components,
+                        };
+
+                        if !self.wrapped.math.insert(wrapped) {
+                            continue;
+                        }
+
+                        self.write_value_type(module, arg_ty)?;
+                        write!(self.out, " {ABS_FUNCTION}(")?;
+                        self.write_value_type(module, arg_ty)?;
+                        writeln!(self.out, " val) {{")?;
+
+                        let level = crate::back::Level(1);
+                        writeln!(
+                            self.out,
+                            "{level}return val >= 0 ? val : asint(~asuint(val) + 1u);"
+                        )?;
+                        writeln!(self.out, "}}")?;
+                        writeln!(self.out)?;
+                    }
+                    _ => {}
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(super) fn write_wrapped_unary_ops(
+        &mut self,
+        module: &crate::Module,
+        func_ctx: &FunctionCtx,
+    ) -> BackendResult {
+        for (_, expression) in func_ctx.expressions.iter() {
+            if let crate::Expression::Unary { op, expr } = *expression {
+                let expr_ty = func_ctx.resolve_type(expr, &module.types);
+                let (scalar, vector_size) = match *expr_ty {
+                    crate::TypeInner::Scalar(scalar) => (scalar, None),
+                    crate::TypeInner::Vector { size, scalar } => (scalar, Some(size)),
+                    _ => continue,
+                };
```

Comment:
You repeat this match a lot. Pity the poor reader who has to check that these are all the same, and wonder why if they're not. How about turning it into an `impl TypeInner` utility function (`vector_size_and_scalar`, say) in `naga/src/proc/mod.rs`?

Reply:
done. the helpers in `proc/mod.rs` seem to have been recently moved to `proc/type_methods.rs` so that's where I put it